Feature/extended simcom support (IDFGH-7243)

Sjurinho opened the pull request:
## Changes
- Added lists to esp_modem_command_declare.inc
- Extended command library to include many commonly used AT commands (using simcom modules as a base)
- Added SIM7070, SIM7000 and extended SIM7600 classes
  - Tested both C and C++ APIs
- Changed initial modem_mode to be set to UNDEF
  - In case of mcu restart without modem restart, which could make the modem be stuck in data mode

david-cermak commented on components/esp_modem/src/esp_modem_command_library.cpp:
```diff
@@ -18,6 +18,7 @@
 #include "cxx_include/esp_modem_dte.hpp"
 #include "cxx_include/esp_modem_dce_module.hpp"
 #include "cxx_include/esp_modem_command_library.hpp"
+#include <iomanip>
```
Nitpick: Please move up, next to the standard lib `#include <>`'s

david-cermak commented on components/esp_modem/src/esp_modem_command_library.cpp:
```diff
@@ -399,4 +406,138 @@ command_result get_signal_quality(CommandableIf *t, int &rssi, int &ber)
+        // OR-operation to add bands
+        band_bits |= 1 << bands[i];
+    }
+    std::stringstream stream;
```
Could you please check if using the `stringstream` doesn't have any big influence on flash size or static memory usage? I recall we had some issues in the past when using C++ *streams* on ESP platform.

david-cermak replied:
I would just check the static sizes using [idf.py size/size-files](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/performance/size.html#size-summary-idf-py-size) before and after your changes (i.e. using the stream)
I don't think the actual creation of the `stringstream` would have a dramatic impact on heap (was just concerned about the statics)

diplfranzhoepfinger replied:
hi, i can confirm that the used Flash size siginificantly increase. 
before i got no Message about Partition too small, now i get one.

diplfranzhoepfinger replied:
![image](https://user-images.githubusercontent.com/37779037/168086783-5f132704-8940-4620-a00e-8a3fc5a201da.png)

diplfranzhoepfinger replied:
before: 

![image](https://user-images.githubusercontent.com/37779037/168087137-4878dadd-c32c-4074-9a97-d34ff2a0e8c8.png)

david-cermak commented on components/esp_modem/include/generate/esp_modem_command_declare.inc:
```diff
@@ -36,6 +37,7 @@
 #define BOOL_IN(param, name) const bool _ARG(param, name)
 #define BOOL_OUT(param, name) bool* _ARG(param, name)
 #define INT_OUT(param, name) int* _ARG(param, name)
+#define LIST_IN(param, name) const int* _ARG(param, name)
```
Could we make the `LIST_IN` type array instead of ptr? (we should also call in *integer-list* or similar...)

Sjurinho replied:
We probably could, I just wasn't sure exactly how to navigate the _ARG definitions. Although I think it might be better to use pointers and rather have _ARG take in pointer + size?

david-cermak replied:
Would this work?
```suggestion
#define LIST_IN(param, name) const int _ARG(param[], name[])
```
it still uses pointer and sizes, but is syntactically closer to an array.

but very minor point, feel free to resolve.

Sjurinho replied:
If we want to make it more general, perhaps we could make it a `const void*` ?

david-cermak replied:
> If we want to make it more general, perhaps we could make it a const void* ? 

I personally prefer concrete types, so the list of integers looks good to me. Resolving, as I'm finding this okay.

Just to explain why I commented:
1) Types `int *` and `int []` are *almost* the same, but the latter suggest that it's an array (while the former is just a pointer)
2) Naming it `LIST_IN()` may suggest any kind of list, but it is a list of **integers** per your definition.

(Optionally you can just rename it to `INT_LIST_IN()` if that sounds good to you)

david-cermak reviewed (approved):
Nice work, thanks! I don't have the devices to test your changes, but the PR looks very good to me!

david-cermak commented:
Please check the unit test failure in https://github.com/espressif/esp-protocols/runs/6170758917?check_suite_focus=true

Sjurinho commented:
@david-cermak Could you elaborate where the host test failed? 
I see that the last console print was during the CMUX test, but as far as I can see I haven't altered the CMUX part of the code anywhere, except for changing the initial mode to be `UNDEF`

david-cermak commented:
> @david-cermak Could you elaborate on where the host test failed?

I don't see much in the logs either. I'm sorry, will have to improve the error reporting in the CI!

Running it locally makes it clear:
```
-------------------------------------------------------------------------------
DCE modes
-------------------------------------------------------------------------------
../main/test_modem.cpp:120
...............................................................................

../main/test_modem.cpp:131: FAILED:
  CHECK( dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == false )
with expansion:
  true == false
```

so your assumption about the initial state was correct. This change makes the unit test pass:
```patch
diff --git a/components/esp_modem/test/host_test/main/test_modem.cpp b/components/esp_modem/test/host_test/main/test_modem.cpp
index f2305e9..ff46808 100644
--- a/components/esp_modem/test/host_test/main/test_modem.cpp
+++ b/components/esp_modem/test/host_test/main/test_modem.cpp
@@ -128,7 +128,7 @@ TEST_CASE("DCE modes", "[esp_modem]")
     auto dce = create_SIM7600_dce(&dce_config, dte, &netif);
     CHECK(dce != nullptr);
 
-    CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == false);
+    CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == true);
     CHECK(dce->set_mode(esp_modem::modem_mode::DATA_MODE) == true);
     CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == true);
 }
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-protocols?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-protocols?pullRequest=30) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-protocols?pullRequest=30) it.</sub>

diplfranzhoepfinger commented:
i want to test this ! i have a SIM7070 !

cj8scrambler reviewed (commented):
I was able to verify this on a SIM7000G (LilyGo board).  This solution is much more elegant than the hacks I was trying to wedge in.  Just a few documentation suggestions.

Sjurinho commented:
Hey @cj8scrambler Thanks for the feedback. Haven't had the time to work on this for a while, but hopefully will by the end of this week. Can add the notes, however they are related to the specific modem and I am therefore not sure if it is helpful as they might change depending on what modem one uses.

Sjurinho commented:
> i want to test this ! i have a SIM7070 !

Yeah, that would be great. I have tested it, but don't hesitate to report any errors/unclarities you might stumble upon:)

Sjurinho commented:
> > @david-cermak Could you elaborate on where the host test failed?
> 
> I don't see much in the logs either. I'm sorry, will have to improve the error reporting in the CI!
> 
> Running it locally makes it clear:
> 
> ```
> -------------------------------------------------------------------------------
> DCE modes
> -------------------------------------------------------------------------------
> ../main/test_modem.cpp:120
> ...............................................................................
> 
> ../main/test_modem.cpp:131: FAILED:
>   CHECK( dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == false )
> with expansion:
>   true == false
> ```
> 
> so your assumption about the initial state was correct. This change makes the unit test pass:
> 
> ```diff
> diff --git a/components/esp_modem/test/host_test/main/test_modem.cpp b/components/esp_modem/test/host_test/main/test_modem.cpp
> index f2305e9..ff46808 100644
> --- a/components/esp_modem/test/host_test/main/test_modem.cpp
> +++ b/components/esp_modem/test/host_test/main/test_modem.cpp
> @@ -128,7 +128,7 @@ TEST_CASE("DCE modes", "[esp_modem]")
>      auto dce = create_SIM7600_dce(&dce_config, dte, &netif);
>      CHECK(dce != nullptr);
>  
> -    CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == false);
> +    CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == true);
>      CHECK(dce->set_mode(esp_modem::modem_mode::DATA_MODE) == true);
>      CHECK(dce->set_mode(esp_modem::modem_mode::COMMAND_MODE) == true);
>  }
> ```

Did you want me to change the tests in this PR, @david-cermak ?

david-cermak commented:
> Did you want me to change the tests in this PR, @david-cermak ?

Yes, please. Would be great If you can make it in a separate commit. The reason is that you've actually changed the state-machine, i.e. that the modem is in a different state after init. This change makes perfect sense, but we'd have to update the basic behavioral tests, too.

diplfranzhoepfinger commented on components/esp_modem/include/generate/esp_modem_command_declare.inc:
```diff
@@ -200,7 +202,93 @@ ESP_MODEM_DECLARE_DCE_COMMAND(reset, command_result, 0) \
+    \
+/**
+ * @brief Preferred network mode (CAT-M and/or NB-IoT)
+ * @param[in] mode preferred selection
```
this should be really in the Sourcecode or in the Documantation.

diplfranzhoepfinger commented on components/esp_modem/include/generate/esp_modem_command_declare.inc:
```diff
@@ -200,7 +202,93 @@ ESP_MODEM_DECLARE_DCE_COMMAND(reset, command_result, 0) \
+    \
+/**
+ * @brief Set network bands for CAT-M or NB-IoT
+ * @param[in] mode CAT-M or NB-IoT
```
EGSM_MODE
DCS_MODE
ALL_MODE

Sjurinho replied:
Yes, that is correct.

Sjurinho commented:
It took a while, but I've finally come around to update the code. I've removed stringstream and renamed the list-in parameter to be integer_list_in, @david-cermak @diplfranzhoepfinger . I agree with @diplfranzhoepfinger that the documentation should not mention modem-specific comments, so I've removed those, and will not be adding it. Thank you to everyone who has helped test the implementation

david-cermak commented:
@Sjurinho Thanks for the updates! Could you please click though the [CLA assistant](https://github.com/espressif/esp-protocols/pull/30#issuecomment-1111176354)? Need your agreement to accept the PR.

Sjurinho commented:
I believe I have already, @david-cermak

david-cermak commented:
> Can we get the package version bumped so this feature is available?

Good point, thanks. we still have some CMUX related issues, so let me quickly fix those and I'll bump the version after that.

>  Or is there some way to tell idf to pull this package from HEAD?

You can use
* `^1.0.0` notation to get the latest package with `MAJOR_VERSION==1`
* `~0.1.0` notation to get the latest package starting with `0.1`

franz-ms-muc commented:
can you tell a Status about the "some CMUX related issues" ?

david-cermak commented:
> can you tell a Status about the "some CMUX related issues" ?

turns out to be multiple related issues, that would require some minor refactor.

I would try to complete at least this https://github.com/espressif/esp-protocols/pull/53 simple fix and will update the version. Will do by the end of this week.
